# Cross-Validation and the Bootstrap

<a target="_blank" href="https://colab.research.google.com/github/intro-stat-learning/ISLP_labs/blob/v2.2/Ch05-resample-lab.ipynb">
<img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/intro-stat-learning/ISLP_labs/v2.2?labpath=Ch05-resample-lab.ipynb)

In this lab, we explore the resampling techniques covered in this
chapter. Some of the commands in this lab may take a while to run on
your computer.

We again begin by placing most of our imports at this top level.

In [2]:
!pip install ISLP
import numpy as np
import statsmodels.api as sm
from ISLP import load_data
from ISLP.models import (ModelSpec as MS,
                         summarize,
                         poly)
from sklearn.model_selection import train_test_split


There are several new imports needed for this lab.

In [3]:
from functools import partial
from sklearn.model_selection import \
     (cross_validate,
      KFold,
      ShuffleSplit)
from sklearn.base import clone
from ISLP.models import sklearn_sm


## The Validation Set Approach
We explore the use of the validation set approach in order to estimate
the test error rates that result from fitting various linear models on
the  `Auto`  data set. https://islp.readthedocs.io/en/latest/datasets/Auto.html

![image.png](attachment:image.png)

We use the function `train_test_split()` to split
the data into training and validation sets. As there are 392 observations,
we split into two equal sets of size 196 using the
argument `test_size=196`. It is generally a good idea to set a random seed
when performing operations like this that contain an
element of randomness, so that the results obtained can be reproduced
precisely at a later time. We set the random seed of the splitter
with the argument `random_state=0`.

In [4]:
Auto = load_data('Auto')
print(Auto.info)

Auto_train, Auto_valid = train_test_split(Auto,
                                         test_size=196,
                                         random_state=0)

Auto_valid.describe


<bound method DataFrame.info of                             mpg  cylinders  displacement  horsepower  weight  \
name                                                                           
chevrolet chevelle malibu  18.0          8         307.0         130    3504   
buick skylark 320          15.0          8         350.0         165    3693   
plymouth satellite         18.0          8         318.0         150    3436   
amc rebel sst              16.0          8         304.0         150    3433   
ford torino                17.0          8         302.0         140    3449   
...                         ...        ...           ...         ...     ...   
ford mustang gl            27.0          4         140.0          86    2790   
vw pickup                  44.0          4          97.0          52    2130   
dodge rampage              32.0          4         135.0          84    2295   
ford ranger                28.0          4         120.0          79    2625   
chevy s-

<bound method NDFrame.describe of                              mpg  cylinders  displacement  horsepower  weight  \
name                                                                            
dodge colt                  28.0          4          90.0          75    2125   
ford fairmont 4             22.3          4         140.0          88    2890   
oldsmobile delta 88 royale  12.0          8         350.0         160    4456   
plymouth horizon miser      38.0          4         105.0          63    2125   
subaru dl                   33.8          4          97.0          67    2145   
...                          ...        ...           ...         ...     ...   
toyota celica gt liftback   21.1          4         134.0          95    2515   
audi 5000                   20.3          5         131.0         103    2830   
chevrolet concours          17.5          6         250.0         110    3520   
ford torino 500             19.0          6         250.0          88    3302   
amc concord d/l             18.1          6         258.0         120    3410   

                            acceleration  year  origin  
name                                                    
dodge colt                          14.5    74       1  
ford fairmont 4                     17.3    79       1  
oldsmobile delta 88 royale          13.5    72       1  
plymouth horizon miser              14.7    82       1  
subaru dl                           18.0    80       3  
...                                  ...   ...     ...  
toyota celica gt liftback           14.8    78       3  
audi 5000                           15.9    78       2  
chevrolet concours                  16.4    77       1  
ford torino 500                     15.5    71       1  
amc concord d/l                     15.1    78       1  

[196 rows x 8 columns]>

Now we can fit a linear regression using only the observations corresponding to the training set `Auto_train`.

In [5]:
hp_mm = MS(['horsepower'])
X_train = hp_mm.fit_transform(Auto_train)
y_train = Auto_train['mpg']
model = sm.OLS(y_train, X_train)
results = model.fit()


We now use the `predict()` method of `results` evaluated on the model matrix for this model
created using the validation data set. We also calculate the validation MSE of our model.

In [6]:
X_valid = hp_mm.transform(Auto_valid)
y_valid = Auto_valid['mpg']
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2) # MSE on validation dataset


np.float64(23.61661706966988)

Hence our estimate for the validation MSE of  the linear regression
fit is $23.62$.

We can also estimate the validation error for
higher-degree polynomial regressions. We first provide a function `evalMSE()` that takes a model string as well
as a training and test set and returns the MSE on the test set.

In [7]:
def evalMSE(terms,
            response,
            train,
            test):

   mm = MS(terms)
   X_train = mm.fit_transform(train)
   y_train = train[response]

   X_test = mm.transform(test)
   y_test = test[response]

   results = sm.OLS(y_train, X_train).fit()
   test_pred = results.predict(X_test)

   return np.mean((y_test - test_pred)**2)


Let’s use this function to estimate the validation MSE
using linear, quadratic and cubic fits. We use the `enumerate()`  function
here, which gives both the values and indices of objects as one iterates
over a for loop.

In [31]:
#help(enumerate)

MSE = np.zeros(3)
for idx, degree in enumerate(range(1, 4)):
    MSE[idx] = evalMSE([poly('horsepower', degree)],
                       'mpg',
                       Auto_train,
                       Auto_valid)
MSE


array([20.75540796, 16.94510676, 16.97437833])

These error rates are $23.62, 18.76$, and $18.80$, respectively. If we
choose a different training/validation split instead, then we
can expect somewhat different errors on the validation set.

In [9]:
Auto_train, Auto_valid = train_test_split(Auto,
                                          test_size=196,
                                          random_state=3)
MSE = np.zeros(3)
for idx, degree in enumerate(range(1, 4)):
    MSE[idx] = evalMSE([poly('horsepower', degree)],
                       'mpg',
                       Auto_train,
                       Auto_valid)
MSE

array([20.75540796, 16.94510676, 16.97437833])

Using this split of the observations into a training set and a validation set,
we find that the validation set error rates for the models with linear, quadratic, and cubic terms are $20.76$, $16.95$, and $16.97$, respectively.

These results are consistent with our previous findings: a model that
predicts `mpg` using a quadratic function of `horsepower`
performs better than a model that involves only a linear function of
`horsepower`, and there is no evidence of an improvement in using a cubic function of `horsepower`.

## Cross-Validation
In theory, the cross-validation estimate can be computed for any generalized
linear model.  {}
In practice, however, the simplest way to cross-validate in
Python is to use `sklearn`, which has a different interface or API
than `statsmodels`, the code we have been using to fit GLMs.

This is a problem which often confronts data scientists: "I have a function to do task $A$, and need to feed it into something that performs task $B$, so that I can compute $B(A(D))$, where $D$ is my data." When $A$ and $B$ don’t naturally speak to each other, this
requires the use of a *wrapper*.
In the `ISLP` package,
we provide
a wrapper, `sklearn_sm()`, that enables us to easily use the cross-validation tools of `sklearn` with
models fit by `statsmodels`.

The class `sklearn_sm()`
has  as its first argument
a model from `statsmodels`. It can take two additional
optional arguments: `model_str` which can be
used to specify a formula, and `model_args` which should
be a dictionary of additional arguments used when fitting
the model. For example, to fit a logistic regression model
we have to specify a `family` argument. This
is passed as `model_args={'family':sm.families.Binomial()}`.

Here is our wrapper in action:

In [10]:

hp_model = sklearn_sm(sm.OLS,
                      MS(['horsepower']))
X, Y = Auto.drop(columns=['mpg']), Auto['mpg']

# help(cross_validate)
print(Auto.shape[0])
cv_results = cross_validate(hp_model,
                            X,
                            Y,
                            cv=Auto.shape[0]  # 392 rows (Determines the cross-validation splitting strategy)
                            )
# print(cv_results)
cv_err = np.mean(cv_results['test_score'])
cv_err



392


np.float64(24.23151351792922)

The arguments to `cross_validate()` are as follows: an
object with the appropriate `fit()`, `predict()`,
and `score()` methods,  an
array of features `X` and a response `Y`.
We also included an additional argument `cv` to `cross_validate()`; specifying an integer
$K$ results in $K$-fold cross-validation. We have provided a value
corresponding to the total number of observations, which results in
leave-one-out cross-validation (LOOCV). The `cross_validate()`  function produces a dictionary with several components;
we simply want the cross-validated test score here (MSE), which is estimated to be 24.23.

We can repeat this procedure for increasingly complex polynomial fits.
To automate the process, we again
use a for loop which iteratively fits polynomial
regressions of degree 1 to 5, computes the
associated cross-validation error, and stores it in the $i$th element
of the vector `cv_error`. The variable `d` in the for loop
corresponds to the degree of the polynomial. We begin by initializing the
vector. This command may take a couple of seconds to run.

In [11]:
cv_error = np.zeros(5)
H = np.array(Auto['horsepower'])
M = sklearn_sm(sm.OLS)

# help(np.power.outer)
# Loop over polynomial degrees from 1 to 5 using 'enumerate' to get both the index 'i' and degree 'd'
for i, d in enumerate(range(1,6)):

    # np.power.outer(H, np.arange(d+1)) generates [1, H, H^2, ..., H^d] for each data point
    X = np.power.outer(H, np.arange(d+1))

    # Perform cross-validation using the design matrix X and target variable Y
    M_CV = cross_validate(M,
                          X,
                          Y,
                          cv=Auto.shape[0]) # implies leave-one-out cross-validation, where each sample is used as a test set once

    # Compute the mean of the cross-validation scores and store it in the cv_error array
    cv_error[i] = np.mean(M_CV['test_score'])
cv_error

#print(X)
#print(H)


array([24.23151352, 19.24821312, 19.33498406, 19.42443029, 19.03320648])

In [32]:
X = np.power.outer?

As in Figure~\ref{Ch5:cvplot}, we see a sharp drop in the estimated test MSE between the linear and
quadratic fits, but then no clear improvement from using higher-degree polynomials.

Above we introduced the `outer()`  method of the `np.power()`
function.  The `outer()` method is applied to an operation
that has two arguments, such as `add()`, `min()`, or
`power()`.
It has two arrays as
arguments, and then forms a larger
array where the operation is applied to each pair of elements of the
two arrays.

In [12]:
A = np.array([3, 5, 9])
B = np.array([2, 4])
np.add.outer(A, B)

# help(np.outer)


array([[ 5,  7],
       [ 7,  9],
       [11, 13]])

In the CV example above, we used $K=n$, but of course we can also use $K<n$. The code is very similar
to the above (and is significantly faster). Here we use `KFold()` to partition the data into $K=10$ random groups. We use `random_state` to set a random seed and initialize a vector `cv_error` in which we will store the CV errors corresponding to the
polynomial fits of degrees one to five.

In [13]:
cv_error = np.zeros(5)

# help(KFold)

cv = KFold(n_splits=10,
           shuffle=True,   # Whether to shuffle the data before splitting into batches. Note that the samples within each split will not be shuffled.
           random_state=0) # use same splits for each degree. When `shuffle` is True, `random_state` affects the ordering of the indices, which controls the randomness of each fold. Otherwise, this parameter has no effect

# print(cv)

for i, d in enumerate(range(1,6)):

    X = np.power.outer(H, np.arange(d+1))

    M_CV = cross_validate(M,
                          X,
                          Y,
                          cv=cv)

    cv_error[i] = np.mean(M_CV['test_score'])

cv_error


array([24.20766449, 19.18533142, 19.27626666, 19.47848403, 19.13720065])

Notice that the computation time is much shorter than that of LOOCV.
(In principle, the computation time for LOOCV for a least squares
linear model should be faster than for $K$-fold CV, due to the
availability of the formula~(\ref{Ch5:eq:LOOCVform})  for LOOCV;
however, the generic `cross_validate()`  function does not make
use of this formula.)  We still see little evidence that using cubic
or higher-degree polynomial terms leads to a lower test error than simply
using a quadratic fit.

The `cross_validate()` function is flexible and can take
different splitting mechanisms as an argument. For instance, one can use the `ShuffleSplit()` funtion to implement
the validation set approach just as easily as K-fold cross-validation.

In [14]:
# help(ShuffleSplit)

validation = ShuffleSplit(n_splits=1,
                          test_size=196,
                          random_state=0)

results = cross_validate(hp_model,
                         Auto.drop(['mpg'], axis=1),
                         Auto['mpg'],
                         cv=validation);

results['test_score']


array([23.61661707])

https://amueller.github.io/aml/04-model-evaluation/1-data-splitting-strategies.html

![image-2.png](attachment:image-2.png)
![image.png](attachment:image.png)

One can estimate the variability in the test error by running the following:

In [15]:
validation = ShuffleSplit(n_splits=10,
                          test_size=196,
                          random_state=0)
results = cross_validate(hp_model,
                         Auto.drop(['mpg'], axis=1),
                         Auto['mpg'],
                         cv=validation)
results['test_score'].mean(), results['test_score'].std()


(np.float64(23.802232661034168), np.float64(1.4218450941091842))

Note that this standard deviation is not a valid estimate of the
sampling variability of the mean test score or the individual scores, since the randomly-selected training
samples overlap and hence introduce correlations. But it does give an
idea of the Monte Carlo variation
incurred by picking different random folds.

## The Bootstrap
We illustrate the use of the bootstrap in the simple example
 {of Section~\ref{Ch5:sec:bootstrap},}  as well as on an example involving
estimating the accuracy of the linear regression model on the  `Auto`
data set.
### Estimating the Accuracy of a Statistic of Interest
One of the great advantages of the bootstrap approach is that it can
be applied in almost all situations. No complicated mathematical
calculations are required. While there are several implementations
of the bootstrap in Python, its use for estimating
standard error is simple enough that we write our own function
below for the case when our data is stored
in a dataframe.

https://islp.readthedocs.io/en/latest/datasets/Portfolio.html
![image.png](attachment:image.png)

To illustrate the bootstrap, we
start with a simple example.
The  `Portfolio`  data set in the `ISLP` package is described
in Section~\ref{Ch5:sec:bootstrap}. The goal is to estimate the
sampling variance of the parameter $\alpha$ given in formula~(\ref{Ch5:min.var}).  We will
create a function
`alpha_func()`, which takes as input a dataframe `D` assumed
to have columns `X` and `Y`, as well as a
vector `idx` indicating which observations should be used to
estimate
$\alpha$. The function then outputs the estimate for $\alpha$ based on
the selected observations.

In [16]:
Portfolio = load_data('Portfolio')

def alpha_func(D, idx):
   cov_ = np.cov(D[['X','Y']].loc[idx], rowvar=False)
   #print(cov_)
   return ((cov_[1,1] - cov_[0,1]) /
           (cov_[0,0]+cov_[1,1]-2*cov_[0,1]))


def alpha_func(D, idx):
    # Calculate the covariance matrix of columns 'X' and 'Y' from DataFrame D, using only the rows specified by the indices in idx.
    # 'rowvar=False' ensures that each column is treated as a variable, and rows correspond to observations.
    cov_ = np.cov(D[['X', 'Y']].loc[idx], rowvar=False)

    # Return the calculated alpha value, which measures the difference between the variances of 'X' and 'Y' relative to their covariance.
    # The formula is: (Variance of Y - Covariance between X and Y) / (Variance of X + Variance of Y - 2 * Covariance between X and Y)
    return ((cov_[1, 1] - cov_[0, 1]) /
            (cov_[0, 0] + cov_[1, 1] - 2 * cov_[0, 1]))



This function returns an estimate for $\alpha$
based on applying the minimum
    variance formula (\ref{Ch5:min.var}) to the observations indexed by
the argument `idx`.  For instance, the following command
estimates $\alpha$ using all 100 observations.

In [17]:
alpha_func(Portfolio, range(100))

# print((1.30823747-0.62635829)/(1.12864236+1.30823747-2*0.62635829))

np.float64(0.57583207459283)

Next we randomly select
100 observations from `range(100)`, with replacement. This is equivalent
to constructing a new bootstrap data set and recomputing $\hat{\alpha}$
based on the new data set.

In [18]:
# help(np.random.default_rng)
# rng.choice(100,100)
rng= np.random.default_rng(0)
print(rng)

alpha_func(
    Portfolio,
    rng.choice(
        100,       # Randomly select from numbers between 0 and 99 (inclusive)
        100,       # Draw 100 samples
        replace=True  # Allow drawing the same number more than once (sampling with replacement)
    )
)


Generator(PCG64)


np.float64(0.6074452469619004)

This process can be generalized to create a simple function `boot_SE()` for
computing the bootstrap standard error for arbitrary
functions that take only a data frame as an argument.

In [19]:
def boot_SE(func,
            D,
            n=None,
            B=1000,
            seed=0):

    # Initialize the random number generator with a specified seed for reproducibility and other accmulators
    rng = np.random.default_rng(seed)
    first_, second_ = 0, 0

    # Set 'n' to the number of rows in D if 'n' is not provided
    n = n or D.shape[0]

    # Perform bootstrapping B times
    for _ in range(B):

        # Randomly select 'n' indices from the DataFrame D, with replacement
        idx = rng.choice(D.index,
                         n,
                         replace=True)

        # Apply the provided function 'func' to the DataFrame D, using the sampled indices
        value = func(D, idx)

        # Accumulate the value and the square of the value
        first_ += value
        second_ += value**2

    # Calculate the standard error using the formula for the standard deviation
    return np.sqrt(second_ / B - (first_ / B)**2)

![image.png](attachment:image.png)

Notice the use of `_` as a loop variable in `for _ in range(B)`. This is often used if the value of the counter is
unimportant and simply makes sure  the loop is executed `B` times.

Let’s use our function to evaluate the accuracy of our
estimate of $\alpha$ using $B=1{,}000$ bootstrap replications.

In [20]:
alpha_SE = boot_SE(alpha_func,
                   Portfolio,
                   B=1000,
                   seed=0)
alpha_SE


np.float64(0.09118176521277699)

The final output shows that the bootstrap estimate for ${\rm SE}(\hat{\alpha})$ is $0.0912$.

### Estimating the Accuracy of a Linear Regression Model
The bootstrap approach can be used to assess the variability of the
coefficient estimates and predictions from a statistical learning
method. Here we use the bootstrap approach in order to assess the
variability of the estimates for $\beta_0$ and $\beta_1$, the
intercept and slope terms for the linear regression model that uses
`horsepower` to predict `mpg` in the  `Auto`  data set. We
will compare the estimates obtained using the bootstrap to those
obtained using the formulas for ${\rm SE}(\hat{\beta}_0)$ and
${\rm SE}(\hat{\beta}_1)$ described in Section~\ref{Ch3:secoefsec}.

To use our `boot_SE()` function, we must write a function (its
first argument)
that takes a data frame `D` and indices `idx`
as its only arguments. But here we want to bootstrap a specific
regression model, specified by a model formula and data. We show how
to do this in a few simple steps.

We start by writing a generic
function `boot_OLS()` for bootstrapping a regression model that takes a formula to
define the corresponding regression. We use the `clone()` function to
make a copy of the formula that can be refit to the new dataframe. This means
that any derived features such as those defined by `poly()`
(which we will see shortly),
will be re-fit on the resampled data frame.

In [21]:
def boot_OLS(model_matrix, response, D, idx):
    D_ = D.loc[idx]
    Y_ = D_[response]
    X_ = clone(model_matrix).fit_transform(D_)
    return sm.OLS(Y_, X_).fit().params

This is not quite what is needed as the first argument to
`boot_SE()`. The first two arguments which specify the model will not change in the
bootstrap process, and we would like to *freeze* them.   The
function `partial()` from the `functools` module  does precisely this: it takes a function
as an argument, and freezes some of its arguments, starting from the
left. We use it to freeze the first two model-formula arguments of `boot_OLS()`.

In [22]:
hp_func = partial(boot_OLS, MS(['horsepower']), 'mpg')


Typing `hp_func?` will show that it has two arguments `D`
and `idx` --- it is a version of `boot_OLS()` with the first
two arguments frozen --- and hence is ideal as the first argument for `boot_SE()`.

The `hp_func()` function can now be used in order to create
bootstrap estimates for the intercept and slope terms by randomly
sampling from among the observations with replacement. We first
demonstrate its utility on 10 bootstrap samples.

In [23]:
rng = np.random.default_rng(0)
np.array([hp_func(Auto,
          rng.choice(Auto.index,
                     392,
                     replace=True)) for _ in range(10)])


array([[39.12226577, -0.1555926 ],
       [37.18648613, -0.13915813],
       [37.46989244, -0.14112749],
       [38.56723252, -0.14830116],
       [38.95495707, -0.15315141],
       [39.12563927, -0.15261044],
       [38.45763251, -0.14767251],
       [38.43372587, -0.15019447],
       [37.87581142, -0.1409544 ],
       [37.95949036, -0.1451333 ]])

Next, we use the `boot_SE()` {}  function to compute the standard
errors of 1,000 bootstrap estimates for the intercept and slope terms.

In [24]:
hp_se = boot_SE(hp_func,
                Auto,
                B=1000,
                seed=10)
hp_se


intercept     0.731176
horsepower    0.006092
dtype: float64

This indicates that the bootstrap estimate for ${\rm SE}(\hat{\beta}_0)$ is
0.85, and that the bootstrap
estimate for ${\rm SE}(\hat{\beta}_1)$ is
0.0074.  As discussed in
Section~\ref{Ch3:secoefsec}, standard formulas can be used to compute
the standard errors for the regression coefficients in a linear
model. These can be obtained using the `summarize()`  function
from `ISLP.sm`.

In [25]:
hp_model.fit(Auto, Auto['mpg'])
model_se = summarize(hp_model.results_)['std err']
model_se


intercept     0.717
horsepower    0.006
Name: std err, dtype: float64

The standard error estimates for $\hat{\beta}_0$ and $\hat{\beta}_1$
obtained using the formulas  from Section~\ref{Ch3:secoefsec}  are
0.717 for the
intercept and
0.006 for the
slope. Interestingly, these are somewhat different from the estimates
obtained using the bootstrap.  Does this indicate a problem with the
bootstrap? In fact, it suggests the opposite.  Recall that the
standard formulas given in
 {Equation~\ref{Ch3:se.eqn} on page~\pageref{Ch3:se.eqn}}
rely on certain assumptions. For example,
they depend on the unknown parameter $\sigma^2$, the noise
variance. We then estimate $\sigma^2$ using the RSS. Now although the
formula for the standard errors do not rely on the linear model being
correct, the estimate for $\sigma^2$ does.  We see
 {in Figure~\ref{Ch3:polyplot} on page~\pageref{Ch3:polyplot}}  that there is
a non-linear relationship in the data, and so the residuals from a
linear fit will be inflated, and so will $\hat{\sigma}^2$.  Secondly,
the standard formulas assume (somewhat unrealistically) that the $x_i$
are fixed, and all the variability comes from the variation in the
errors $\epsilon_i$.  The bootstrap approach does not rely on any of
these assumptions, and so it is likely giving a more accurate estimate
of the standard errors of $\hat{\beta}_0$ and $\hat{\beta}_1$ than
the results from `sm.OLS`.

Below we compute the bootstrap standard error estimates and the
standard linear regression estimates that result from fitting the
quadratic model to the data. Since this model provides a good fit to
the data (Figure~\ref{Ch3:polyplot}), there is now a better
correspondence between the bootstrap estimates and the standard
estimates of ${\rm SE}(\hat{\beta}_0)$, ${\rm SE}(\hat{\beta}_1)$ and
${\rm SE}(\hat{\beta}_2)$.

In [26]:
quad_model = MS([poly('horsepower', 2, raw=True)])
quad_func = partial(boot_OLS,
                    quad_model,
                    'mpg')
boot_SE(quad_func, Auto, B=1000)


intercept                                  1.538641
poly(horsepower, degree=2, raw=True)[0]    0.024696
poly(horsepower, degree=2, raw=True)[1]    0.000090
dtype: float64

We  compare the results to the standard errors computed using `sm.OLS()`.

In [27]:
M = sm.OLS(Auto['mpg'],
           quad_model.fit_transform(Auto))
summarize(M.fit())['std err']


intercept                                  1.800
poly(horsepower, degree=2, raw=True)[0]    0.031
poly(horsepower, degree=2, raw=True)[1]    0.000
Name: std err, dtype: float64

# Lab


# Section 5.3.1: Modified Validation Set Approach

Select a random training set (without replacement) of 250 observations out of the original 392 observations in the Auto data set with 'random_state=10'. Fit linear, quadratic, and cubic regressions to the 250 observations in the training set and then calculate the MSE of the other 142 observations in the validation set. Enter the values of estimated test MSE to the linear, quadratic, and cubic regression models (round to two decimal places).

In [35]:
Auto = load_data('Auto')
print(Auto.info)

Auto_train, Auto_valid = train_test_split(Auto,
                                         test_size=142,
                                         random_state=10)

<bound method DataFrame.info of                             mpg  cylinders  displacement  horsepower  weight  \
name                                                                           
chevrolet chevelle malibu  18.0          8         307.0         130    3504   
buick skylark 320          15.0          8         350.0         165    3693   
plymouth satellite         18.0          8         318.0         150    3436   
amc rebel sst              16.0          8         304.0         150    3433   
ford torino                17.0          8         302.0         140    3449   
...                         ...        ...           ...         ...     ...   
ford mustang gl            27.0          4         140.0          86    2790   
vw pickup                  44.0          4          97.0          52    2130   
dodge rampage              32.0          4         135.0          84    2295   
ford ranger                28.0          4         120.0          79    2625   
chevy s-

In [36]:
hp_mm = MS(['horsepower'])
X_train = hp_mm.fit_transform(Auto_train)
y_train = Auto_train['mpg']
model = sm.OLS(y_train, X_train)
results = model.fit()

In [37]:
X_valid = hp_mm.transform(Auto_valid)
y_valid = Auto_valid['mpg']
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2) # MSE on validation dataset

np.float64(23.015471920950535)

In [38]:
def evalMSE(terms,
            response,
            train,
            test):

   mm = MS(terms)
   X_train = mm.fit_transform(train)
   y_train = train[response]

   X_test = mm.transform(test)
   y_test = test[response]

   results = sm.OLS(y_train, X_train).fit()
   test_pred = results.predict(X_test)

   return np.mean((y_test - test_pred)**2)


In [39]:
MSE = np.zeros(3)
for idx, degree in enumerate(range(1, 4)):
    MSE[idx] = evalMSE([poly('horsepower', degree)],
                       'mpg',
                       Auto_train,
                       Auto_valid)
MSE


array([23.01547192, 17.73173205, 17.99325931])

# Section 5.3.2: Leave-One-Out Cross-Validation

What is the leave-one-out cross-validation estimate of the test MSE for the sixth-order polynomial regression model for the Auto data set (using "mpg" as the response variable and "horsepower" as the predictor variable)? Round your answer to 2 decimal places.

In [40]:
hp_model = sklearn_sm(sm.OLS,
                      MS(['horsepower']))
X, Y = Auto.drop(columns=['mpg']), Auto['mpg']

# help(cross_validate)
print(Auto.shape[0])
cv_results = cross_validate(hp_model,
                            X,
                            Y,
                            cv=Auto.shape[0]  # 392 rows (Determines the cross-validation splitting strategy)
                            )
# print(cv_results)
cv_err = np.mean(cv_results['test_score'])
cv_err

392


np.float64(24.23151351792922)

In [42]:
cv_error = np.zeros(6)
H = np.array(Auto['horsepower'])
M = sklearn_sm(sm.OLS)

# help(np.power.outer)
# Loop over polynomial degrees from 1 to 6 using 'enumerate' to get both the index 'i' and degree 'd'
for i, d in enumerate(range(1,7)):

    # np.power.outer(H, np.arange(d+1)) generates [1, H, H^2, ..., H^d] for each data point
    X = np.power.outer(H, np.arange(d+1))

    # Perform cross-validation using the design matrix X and target variable Y
    M_CV = cross_validate(M,
                          X,
                          Y,
                          cv=Auto.shape[0]) # implies leave-one-out cross-validation, where each sample is used as a test set once

    # Compute the mean of the cross-validation scores and store it in the cv_error array
    cv_error[i] = np.mean(M_CV['test_score'])
cv_error

#print(X)
#print(H)

array([24.23151352, 19.24821312, 19.33498406, 19.42443029, 19.03320648,
       19.00693693])

# Section 5.3.3: k-Fold Cross-Validation

Implement 5-fold cross-validation for polynomial regression models of orders one to ten for the Auto data set (using "mpg" as the response variable and "horsepower" as the predictor variable) with "random_state=17". What are the 5-fold cross-validation estimates of the test MSE for the linear (order one) and quadratic (order two) regression models? Round your answers to 2 decimal places.

In [43]:
cv_error = np.zeros(5)

# help(KFold)

cv = KFold(n_splits=5,
           shuffle=True,   # Whether to shuffle the data before splitting into batches. Note that the samples within each split will not be shuffled.
           random_state=17) # use same splits for each degree. When `shuffle` is True, `random_state` affects the ordering of the indices, which controls the randomness of each fold. Otherwise, this parameter has no effect

# print(cv)

for i, d in enumerate(range(1,6)):

    X = np.power.outer(H, np.arange(d+1))

    M_CV = cross_validate(M,
                          X,
                          Y,
                          cv=cv)

    cv_error[i] = np.mean(M_CV['test_score'])

cv_error

array([24.06276551, 19.03438137, 19.01789816, 19.24324267, 18.86845234])

# Section 5.3.4: The Bootstrap - Estimating the Accuracy of a Statistic of Interest

Use "seed=2" to implement a bootstrap analysis of the function "alpha_fn" defined for the Portfolio data set. Use 1000 for the number of bootstrap estimates. What is the bootstrap estimate for the standard error of alpha-hat? Round your answer to 3 decimal places.

In [44]:
Portfolio = load_data('Portfolio')

def alpha_func(D, idx):
   cov_ = np.cov(D[['X','Y']].loc[idx], rowvar=False)
   #print(cov_)
   return ((cov_[1,1] - cov_[0,1]) /
           (cov_[0,0]+cov_[1,1]-2*cov_[0,1]))


def alpha_func(D, idx):
    # Calculate the covariance matrix of columns 'X' and 'Y' from DataFrame D, using only the rows specified by the indices in idx.
    # 'rowvar=False' ensures that each column is treated as a variable, and rows correspond to observations.
    cov_ = np.cov(D[['X', 'Y']].loc[idx], rowvar=False)

    # Return the calculated alpha value, which measures the difference between the variances of 'X' and 'Y' relative to their covariance.
    # The formula is: (Variance of Y - Covariance between X and Y) / (Variance of X + Variance of Y - 2 * Covariance between X and Y)
    return ((cov_[1, 1] - cov_[0, 1]) /
            (cov_[0, 0] + cov_[1, 1] - 2 * cov_[0, 1]))

In [46]:
# Use the existing boot_SE and alpha_func functions with the specified seed and number of bootstrap estimates
alpha_SE_seed2 = boot_SE(alpha_func,
                         Portfolio,
                         B=1000,
                         seed=2)

# Print the bootstrap estimate for the standard error of alpha-hat, rounded to 3 decimal places
print(f"The bootstrap estimate for the standard error of alpha-hat (with seed=2) is: {alpha_SE_seed2:.3f}")

The bootstrap estimate for the standard error of alpha-hat (with seed=2) is: 0.091
